### DISCUSSION TITLE
cpp/basics/byte_alignment

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#56] [Announcements]
# cpp/basics/byte_alignment

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/basics/byte_alignment.html

**Comment by 1604148641** [comment]
struct MyStruct {
 double a;    // 8 个字节
 char b;      // 本来占一个字节，但是接下来的 int 需要起始地址为4的倍数
              //所以这里也会加3字节的padding
 int c;       // 4 个字节
 // 总共:  8 + 4 + 4 = 16
};
为啥这里不按整体8个字节对齐，也就是8+8+8

**Reply by TechManDexiang** [reply]
因为a是8个字节，但是后面的b和c可以填充成一个8个字节

**Reply by yysczl** [reply]
MyStruct 结构体的内存布局是基于各个字段的对齐要求进行的。char 字段后面需要填充 3 字节，以确保 int 字段的对齐，而不是将所有字段都填充到 8 字节。这样设计能够在保证内存对齐的同时，优化内存使用效率。
